lib/pacemaker/pcmk_status.c

```diff
@@ -18,6 +20,246 @@
+        out->err(out, "Could not connect to the CIB: %s",
+                 pcmk_rc_str(rc));
+        return rc;
+    }
+
+    rc = cib->cmds->query(cib, NULL, current_cib,
+                          cib_scope_local | cib_sync_call);
+    rc = pcmk_legacy2rc(rc);
+
+    return rc;
+}
+
+static stonith_t *
+fencing_connect(void)
+{
+    stonith_t *st = stonith_api_new();
+    int rc = pcmk_rc_ok;
+
+    if (st == NULL) {
+        return NULL;
+    } else if (st->state != stonith_disconnected) {
+        st->cmds->remove_notification(st, NULL);
+        st->cmds->disconnect(st);
+        stonith_api_delete(st);
+        return NULL;
+    }
+
+    rc = st->cmds->connect(st, crm_system_name, NULL);
+    if (rc == pcmk_rc_ok) {
+        return st;
+    } else {
+        if (st->state != stonith_disconnected) {
+            st->cmds->remove_notification(st, NULL);
+            st->cmds->disconnect(st);
+        }
+
+        stonith_api_delete(st);
+        return NULL;
+    }
+}
+
+static void
+pacemakerd_event_cb(pcmk_ipc_api_t *pacemakerd_api,
+                    enum pcmk_ipc_event event_type, crm_exit_t status,
+                    void *event_data, void *user_data)
+{
+    pcmk_pacemakerd_api_reply_t *reply = event_data;
+    enum pcmk_pacemakerd_state *state =
+        (enum pcmk_pacemakerd_state *) user_data;
+
+    /* we are just interested in the latest reply */
+    *state = pcmk_pacemakerd_state_invalid;
+
+    switch (event_type) {
+        case pcmk_ipc_event_reply:
+            break;
+
+        default:
+            return;
+    }
+
+    if (status != CRM_EX_OK) {
+        return;
+    }
+
+    if (reply->reply_type == pcmk_pacemakerd_reply_ping &&
+        reply->data.ping.last_good != (time_t) 0 &&
+        reply->data.ping.status == pcmk_rc_ok) {
+        *state = reply->data.ping.state;
+    }
+}
+
+static int
+pacemakerd_status(pcmk__output_t *out, cib_t *cib, bool use_cib_native)
+{
+    int rc = pcmk_rc_ok;
+    pcmk_ipc_api_t *pacemakerd_api = NULL;
+    enum pcmk_pacemakerd_state state = pcmk_pacemakerd_state_invalid;
+
+    if (!use_cib_native) {
+        return rc;
+    }
+
+    if (cib != NULL &&
+        (cib->state == cib_connected_query ||
+         cib->state == cib_connected_command)) {
+        return rc;
+    }
+
+    rc = pcmk_new_ipc_api(&pacemakerd_api, pcmk_ipc_pacemakerd);
+    if (pacemakerd_api == NULL) {
+        out->err(out, "Could not connect to pacemakerd: %s",
+                 pcmk_rc_str(rc));
+        return rc;
+    }
+
+    pcmk_register_ipc_callback(pacemakerd_api, pacemakerd_event_cb, (void *) &state);
+
+    rc = pcmk_connect_ipc(pacemakerd_api, pcmk_ipc_dispatch_sync);
+    if (rc != pcmk_rc_ok) {
+        out->err(out, "Could not connect to pacemakerd: %s",
+                 pcmk_rc_str(rc));
+        pcmk_free_ipc_api(pacemakerd_api);
+        return rc;
+    }
+
+    if (state == pcmk_pacemakerd_state_running || state == pcmk_pacemakerd_state_shutting_down) {
+        rc = pcmk_rc_ok;
+    }
```

Comment:
Isn't rc guaranteed to be pcmk_rc_ok at this point?

Reply:
Yeah, this is probably due to copy and paste from tools/crm_mon.c.  I could either just drop this block, or convert it into something like
```
if (state != pcmk_pacemakerd_state_running && state != pcmk_pacemakerd_state_shutting_down) {
    rc = pcmk_rc_error;
}
```

Reply:
Ah, what happened to pcmk_pacemakerd_api_ping()?

Reply:
I didn't think that would be necessary with pcmk_ipc_dispatch_sync.

Reply:
Yes, the ping is the point of the pacemakerd connection. pacemakerd will reply with the status of the subdaemons. The sync dispatch means the event callback will be called directly from the ping function, without needing a main loop.